### PR TITLE
LN Gateways - The christmas version

### PR DESCRIPTION
**Pull request by okjodom**
A proof-of-concept refactor of the gateway to adopt RPC interfacing between the gateway webserver binary (gatewayd) and it's lightning node

Proves spec proposal in #1143 by:
- [x] Implementing `GatewayLightning` for LND from the proto3 spec. This is packaged as `lnd-rpc` binary
- [x] Repackaging the CLN plugin as `GatewayLightning` from the proto3 spec. This is packaged as `cln-rpc` binary
- [x] Implementing and wiring a common Lightning RPC client within `LnGateway`. This can call into rpc servers for either node implementation

## Illustrations
1. Gateway Pay invoice Flow
Screenshot from 2022-12-30 15-21-28

2. Gateway Receive Payment Flow
Screenshot from 2022-12-30 15-22-52

3. Component: Gateway Lightning RPC Client #1198 
Screenshot from 2023-01-07 01-29-52

4. Component: Gateway Lightning RPC Server (CLN extension service) #1224 
Screenshot from 2023-01-06 11-35-46

## WIP
- [ ] Migrate tests to this new architecture
- [ ] Fully implement LND rpc server (all functions unimplemented)
- [ ] Fully implement CLN rpc server (complete HTLC intercept part)

### CODE REVIEW

**Comment by justinmoon** [comment]
Why have 2 separate binaries for lnd / cln? Why not just have 1 binary?

File: gateway/ln-gateway/src/bin/gatewayd.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+use tracing::error;
+
+/**
+ * Fedimint Gateway Binary
+ *
+ * This binary runs a webserver with an API that can be used by Fedimint clients to request routing of payments
+ * through the Lightning Network. It uses a Gateway Lightning RPC client to communicate with a Lightning node.
+ */
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
```
Comment (justinmoon) [inline gateway/ln-gateway/src/bin/gatewayd.rs:19]: I think this needs to configure `tracing` in order to write any logs, right?

File: .tmuxinator.yml [hunk 2]
```diff
@@ -27,16 +27,20 @@ windows:
           - fg
         - ln1:
           - sleep 5 # wait for bitcoind and federation
-          - gateway-cli generate-config 127.0.0.1:8080 'http://127.0.0.1:8080' $FM_CFG_DIR #generate gateway config
-          - lightningd --dev-fast-gossip --dev-bitcoind-poll=1 --network regtest --bitcoin-rpcuser=bitcoin --bitcoin-rpcpassword=bitcoin --lightning-dir=$FM_LN1_DIR --addr=127.0.0.1:9000 --plugin=$FM_BIN_DIR/ln_gateway --fedimint-cfg=$FM_CFG_DIR &
+          - gateway-cli generate-lnrpc-config $FM_LN1_DIR/.gateway 127.0.0.1:9590 #generate gateway config
+          - lightningd --dev-fast-gossip --dev-bitcoind-poll=1 --network regtest --bitcoin-rpcuser=bitcoin --bitcoin-rpcpassword=bitcoin --lightning-dir=$FM_LN1_DIR --addr=127.0.0.1:9000 --plugin=$FM_BIN_DIR/cln-rpc --gateway-cfg=$FM_LN1_DIR/.gateway &
```
Comment (justinmoon) [inline .tmuxinator.yml:31]: should this be `FM_BIN_DIR/cln_rpc`? (underscore instead of dash)

File: gateway/ln-gateway/src/bin/gatewayd.rs (hunk 1 above)
Comment (okjodom) [reply gateway/ln-gateway/src/bin/gatewayd.rs:19]: Probably needs `#[tracing::instrument]` macro marker

**Comment by okjodom** [comment]
> @justinmoon Re: should this be `FM_BIN_DIR/cln_rpc`? (underscore instead of dash)

Our binary names are a mess. We should either prefer underscores, or dashes. Afaik, we have more dashes than underscores at present, so wanted to go with a dashed name for this new bin

**Comment by elsirion** [comment]
> Also, it doesn't seem ideal to require LND gateways to run 2 binaries. Christian Decker suggested we not run the actual gateway code in the plugin for the [following reasons](https://discord.com/channels/990354215060795454/990354215878688860/1051219154918711386) but I don't think this applies to LND at all.

I think what we want regarding the GW are these components (modulo names being different):
* `gatewayd`: The actual GW business logic, connecting to either `cln-gateway-plugin` or `lnd` via RPC
* `cln-gateway-plugin`: Exposing the RPC interface necessary to intercept HTLCs and make payments. Possibly a subset of the `lnd` RPC interface.
* `gateway-cli`: Management tool for `gatewayd`

**Comment by justinmoon** [comment]
> `cln-gateway-plugin`: Exposing the RPC interface necessary to intercept HTLCs and make payments. Possibly a subset of the lnd RPC interface.

Why should this know how to make payments? Why can't `gatewayd` just have a `ClnRpc` reference and make the  payments itself?

**Comment by justinmoon** [comment]
Also `cln-rpc` is slightly confusing name for plugin because we have a [dependency](https://docs.rs/cln-rpc/0.1.0/cln_rpc/) with same name

**Comment by okjodom** [comment]
> Why have 2 separate binaries for lnd / cln? Why not just have 1 binary that can speak both CLN & LND RPC?

This poses a scaling and upgrade challenge in that binary codebase. By your posposal, it should be able to speak CLN, LND, Eclair and so on. If we needed to bump version in service to a CLN feature upgrade / bug fix, suddenly all other node runners are impacted

In contrast, multiple binaries with known RPC interface (defined as proto spec as `GatewayLightning`) mean someone could cobble together a binary of their own, maybe it is a plugin, and run it on their node. Then connect this to our formal `gatewayd` to serve a federation

**Comment by okjodom** [comment]
> Also `cln-rpc` is slightly confusing name for plugin because we have a [dependency](https://docs.rs/cln-rpc/0.1.0/cln_rpc/) with same name

ref #1224 . I like `cln-gateway-plugin` name @elsirion has just proposed. Alternative name could be `gateway-cln-extension`? v `gateway-lnd-extension` etc

**Comment by okjodom** [comment]
> > Also, it doesn't seem ideal to require LND gateways to run 2 binaries. Christian Decker suggested we not run the actual gateway code in the plugin for the [following reasons](https://discord.com/channels/990354215060795454/990354215878688860/1051219154918711386) but I don't think this applies to LND at all.
> 
> I think what we want regarding the GW are these components (modulo names being different):
> 
>     * `gatewayd`: The actual GW business logic, connecting to either `cln-gateway-plugin` or `lnd` via RPC
> 
>     * `cln-gateway-plugin`: Exposing the RPC interface necessary to intercept HTLCs and make payments. Possibly a subset of the `lnd` RPC interface.
> 
>     * `gateway-cli`: Management tool for `gatewayd`

If my current proposal flies, LND gateway operators need to think about three binaries: `gatewayd`, `gateway-lnd-extension` and wll-known `lnd` binaries. They need to think about two binaries at minimum, since LND does'nt have a direct plugin model, thus require  extensions with rpc interfaces into `lnd`.

Justification of a `gatewayd` separate from `gateway-lnd-extension` or `gateway-cln-extension` is so `gatewayd` can remain node implementation agnostic.
In CLN case, we get some resiliency benefits [highlighted by this reasoning](https://discord.com/channels/990354215060795454/990354215878688860/1051219154918711386)

**Comment by justinmoon** [comment]
> This poses a scaling and upgrade challenge in that binary codebase. By your posposal, it should be able to speak CLN, LND, Eclair and so on. If we needed to bump version in service to a CLN feature upgrade / bug fix, suddenly all other node runners are impacted

If a CLN version is fixed and released but a given user is running LND I think they can just ignore the release, right?

**Comment by justinmoon** [comment]
> In contrast, multiple binaries with known RPC interface (defined as proto spec as GatewayLightning) mean someone could cobble together a binary of their own, maybe it is a plugin, and run it on their node. Then connect this to our formal gatewayd to serve a federation

I'm also not sure this will scenario will arise very often. New lightning node implementations (which we wouldn't have support for) don't show up very often.

**Comment by justinmoon** [comment]
Another thing your proposal supports which mine doesn't is having a `gatewayd` instance using CLN & LND at the same time! Also you could have different versions of a node running concurrently by using different versions of `gateway-lnd-extension`. That would be very useful for upgrading multi-node setups with no downtime.

**Comment by okjodom** [comment]
Idea: @justinmoon, @elsirion, perhaps this design naturally opens up a progression for `gatewayd` to move closer to consensus. As a future plugin, could it implement like fulfillment of bolt 12 offers?

For now, we can focus on the concerns raised about having too many binaries, and in my opinion we should mitigate those via mechanisms like
- docker orchestration for `gatewayd` and `gateway-<ln> extension`
- easy one-click install of `gatewayd` for fedimint guardians who want to operate gateways, so they only think about `gateway-<ln> extension` and beyond

cc @dpc, @m1sterc001guy, @jkitman . The discussion thread [staring here](https://github.com/fedimint/fedimint/pull/1168#issuecomment-1373971088) might help build full context

**Comment by justinmoon** [comment]
> perhaps this design naturally opens up a progression for gatewayd to move closer to consensus. As a future plugin, could it implement new features like fulfillment of bolt 12 offers with some sort of custody scheme backed by federation consensus?

My view of things is that we want to have as little running in consensus as possible. `ln` module is that, and `gateway` is the lightning node software that interacts with that module.

For BOLT12 (https://github.com/fedimint/fedimint/issues/559), the `ln` module would need a way to create or upload a BOLT12 offer and have threshold-encrypted preimages uploaded to it. It would also need a way to schnorr sign invoices (assuming federation signs invoice and [not gateway](https://github.com/fedimint/fedimint/issues/559#issuecomment-1374911154)). IIRC BOLT12 invoices are fetched from lightning nodes via lightning p2p network, so the gateway would need a way to intercept these p2p messages (like we do with HTLCs currently) and fetch invoice from federation if needed. But I don't think there's any benefit of having the gateway run as a federation module if it doesn't require consensus. When it does need consensus, just communicate with the `ln` module via its API.

**Comment by justinmoon** [comment]
Some more quesitons:

- Will the extension save anything to disk?
- What happens if `extension` restarts?
- What happens if `gatewayd` restarts?

**Comment by okjodom** [comment]
Thanks for the early reviews and discussions on this POC. Since we agreed on taking this design further, I will cherry pick useful parts and make atomic PRs

**Pull request closed without merging**